### PR TITLE
ARROW-16553: [CI][Java] Adding Java nightly packages (.pom/.jar) to nightlies.apache repository

### PR DESCRIPTION
**Pull request by davisusanibar**
Configure Java to upload artifacts to https://nightlies.apache.org/arrow/java/

### CODE REVIEW

**Comment by assignUser** [comment]
Have you tested this on a fork/test-repo?

File: .github/workflows/java_nightly.yml [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          TMP_FOLDER=arrow_java_$(date +%Y-%m-%d)
+          mkdir -p $TMP_FOLDER
+          pushd $TMP_FOLDER
+          # this download artifacts should be executed thru crossbow (workaround to solve problem about rate limit using github3 without API_TOKEN)
+          echo "**************** 1 - Download $ARROW_JAVA_NIGHTLY_VERSION dependencies ****************"
+          wget $( \
+            wget \
+              "-qO-" "https://api.github.com/repos/ursacomputing/crossbow/releases/tags/$ARROW_JAVA_NIGHTLY_VERSION" \
+              | jq -r '.assets[] | select((.name | endswith(".pom")) or (.name | endswith(".jar"))) | .browser_download_url' \
+              | grep $DEPENDENCY_TO_INSTALL )
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml:98]: This download artifacts should be executed thru crossbow (this is a workaround to solve problem about rate limit using github3 without API_TOKEN):

`github3.exceptions.ForbiddenError: 403 API rate limit exceeded for IP***.154. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)`

Error message: https://github.com/davisusanibar/testing-github-ci/runs/6775682914?check_suite_focus=true

[the remark this answers is not in the document]
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml]: Changed to use archery crossbow download-artifacts and use env CROSSBOW_GITHUB_TOKEN: ${{ github.token }}

[the remark this answers is not in the document]
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml]: This is the log to check the test output: https://github.com/davisusanibar/testing-github-ci/actions/runs/2456672900

Only burnett01/rsync-deployments@5.2 is failing for secrets pending

File: .github/workflows/java_nightly.yml [hunk 2]
```diff
@@ -164,17 +134,18 @@ jobs:
...
-          echo "**************** 3 - Local repository to sync ****************"
           rm -rf repo
-          mkdir repo
-          cp -r ~/.m2/repository/org/apache/arrow repo
-          tree repo
+          mkdir -p repo/org/apache
+          if [ -d "$HOME/.m2/repository/org/apache/arrow" ]; then
+            cp -r ~/.m2/repository/org/apache/arrow repo/org/apache
+          fi
           echo "Repo configured with Arrow Java artifacts ready to sync"
```
Comment (assignUser) [inline .github/workflows/java_nightly.yml:145]: I assume you tested locally that this creates a working maven repo?

File: .github/workflows/java_nightly.yml [hunk 3]
```diff
@@ -60,60 +61,29 @@ jobs:
...
-              | grep $DEPENDENCY_TO_INSTALL )
-          if [ -n "$(ls -A *)" ]; then
+          archery crossbow download-artifacts -f java-jars -t binaries nightly-packaging-$PREFIX-0
+          if [ -n "$(ls -A binaries/*/*/)" ]; then
             echo "Found files!"
             echo "GOT_FILES=true" >> $GITHUB_ENV
           else
             echo "No files found. Stopping upload."
             echo "GOT_FILES=false" >> $GITHUB_ENV
           fi
```
Comment (assignUser) [inline .github/workflows/java_nightly.yml:76]: I have changed this behaviour and removed the envvar in the R workflow in favour of explicitly failing the job with `exit 1` if there are no files downloaded. You can then also remove the `if: `s that reference it

**Review by assignUser — CHANGES_REQUESTED** [review]
Looks good now with the use of archery 👍I added some comments, one thing I was wondering: What is the versioning scheme on the java jars? AFAIK they don't get a new version for each daily build right? Is this something we should keep in mind? Or are you ok with only having one nightly build available?

File: .github/workflows/java_nightly.yml (hunk 2 above)
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml:145]: Yes, `mvn install:install-file` create all the directory structure needed locally to sync

File: .github/workflows/java_nightly.yml [hunk 4]
```diff
@@ -60,60 +61,29 @@ jobs:
...
+            PREFIX=$(date +%Y-%m-%d)
+          fi
           PATTERN_TO_GET_LIB_AND_VERSION='([a-z].+)-([0-9].[0-9].[0-9].dev[0-9]+).([a-z]+)'
-          pushd $TMP_FOLDER
-
-          echo "**************** 2 - Install arrow java libraries to local repository ****************"
+          pushd binaries/nightly-packaging-$PREFIX-0/java-jars
           rm -rf ~/.m2/repository/org/apache/arrow
           for LIBRARY in $(ls | grep -E '.jar|.pom' | grep dev); do
             [[ $LIBRARY =~ $PATTERN_TO_GET_LIB_AND_VERSION ]]
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml:76]: Changed logic to upload all files downloaded

File: .github/workflows/java_nightly.yml (hunk 3 above)
Comment (davisusanibar) [reply .github/workflows/java_nightly.yml]: Changed

File: .github/workflows/java_nightly.yml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        shell: bash
+        run: pip install -e arrow/dev/archery[all]
+      - run: mkdir -p binaries
+      - name: Download Artifacts
+        run: |
+          if [ -z $PREFIX ]; then
+            PREFIX=nightly-packaging-$(date +%Y-%m-%d)-0
+          fi
+          echo $PREFIX
+          archery crossbow download-artifacts -f java-jars -t binaries --skip-pattern-validation $PREFIX
```
Comment (assignUser) [inline .github/workflows/java_nightly.yml:66]:
```suggestion
          archery crossbow download-artifacts -f java-jars -t binaries  $PREFIX
```
The pattern validation will make sure all files are downloaded, is there a reason you want  to skip it?

**Review by assignUser — APPROVED** [review]
@kszucs could you give this a final look over?

**Review by kszucs — APPROVED** [review]
Great! Thanks @davisusanibar! 

I'm going to merge it then trigger a build manually to see everything works as expected.

**Comment by kszucs** [comment]
Triggered build: https://github.com/apache/arrow/actions/runs/2468794467

**Comment by assignUser** [comment]
looks good 🥳 @davisusanibar try https://nightlies.apache.org/arrow/java/ as a maven repo :D

**Comment by assignUser** [comment]
@kszucs Infra did not mention any limits and there is nothing documented I could find.
With R I am planning to keep 14 versions available, with javas much larger overall size a shorter history might be better. I'll check with INFRA.

**Comment by assignUser** [comment]
Not strict limits ~14 versions is fine: https://issues.apache.org/jira/browse/INFRA-23360

**Pull request merged**